ee/clickhouse/queries/cohort_query.py

```diff
@@ -0,0 +1,573 @@
+        self.params.update(condition_params)
+
+        subq = []
+
+        behavior_subquery, behavior_subquery_params, behavior_query_alias = self._get_behavior_subquery()
+        subq.append((behavior_subquery, behavior_query_alias))
+        self.params.update(behavior_subquery_params)
+
+        person_query, person_params, person_query_alias = self._get_persons_query()
+        subq.append((person_query, person_query_alias))
+        self.params.update(person_params)
+
+        sequence_query, sequence_params, sequence_query_alias = self._get_sequence_query()
+        subq.append((sequence_query, sequence_query_alias))
+        self.params.update(sequence_params)
+
+        # Since we can FULL OUTER JOIN, we may end up with pairs of uuids where one side is blank. Always try to choose the non blank ID
+        q, fields = self._build_sources(subq)
+
+        final_query = f"""
+        SELECT {fields} AS id FROM
+        {q}
+        WHERE 1 = 1
+        {conditions}
+        """
+
+        return final_query, self.params
+
+    def _build_sources(self, subq: List[Tuple[str, str]]) -> Tuple[str, str]:
+        q = ""
+        filtered_queries = [(q, alias) for (q, alias) in subq if q and len(q)]
+
+        prev_alias: Optional[str] = None
+        fields = ""
+        for idx, (subq_query, subq_alias) in enumerate(filtered_queries):
+            if idx == 0:
+                q += f"({subq_query}) {subq_alias}"
+                fields = f"{subq_alias}.person_id"
+            elif prev_alias:  # can't join without a previous alias
+                q = f"{q} {full_outer_join_query(subq_query, subq_alias, f'{subq_alias}.person_id', f'{prev_alias}.person_id')}"
+                fields = if_condition(
+                    f"{prev_alias}.person_id = '00000000-0000-0000-0000-000000000000'",
+                    f"{subq_alias}.person_id",
+                    f"{fields}",
+                )
+            prev_alias = subq_alias
+
+        return q, fields
+
+    def _get_behavior_subquery(self) -> Tuple[str, Dict[str, Any], str]:
+        #
+        # Get the subquery for the cohort query.
+        #
+
+        query, params = "", {}
+        if self._should_join_behavioral_query:
+
+            _fields = [f"{self.DISTINCT_ID_TABLE_ALIAS}.person_id AS person_id"]
+            _fields.extend(self._fields)
+
+            date_condition, date_params = self._get_date_condition()
+
+            query = f"""
+            SELECT {", ".join(_fields)} FROM events {self.EVENT_TABLE_ALIAS}
+            {self._get_distinct_id_query()}
+            WHERE team_id = %(team_id)s
+            AND event IN %(events)s
+            {date_condition}
+            GROUP BY person_id
+            """
+
+            query, params = (query, {"team_id": self._team_id, "events": self._events, **date_params,})
+
+        return query, params, self.BEHAVIOR_QUERY_ALIAS
+
+    def _get_persons_query(self) -> Tuple[str, Dict[str, Any], str]:
+        query, params = "", {}
```

Comment:
For `restarted_performing_event` and `performed_event_first_time`, the earliest date is "all time", so we should support that concept here. Makes me wonder if the earliest date should be event specific. Might be overoptimizing, so just a thought.